group-coordinator/src/main/java/org/apache/kafka/coordinator/group/GroupMetadataManager.java

```diff
@@ -3500,6 +3501,42 @@ public void maybeDeleteGroup(String groupId, List<Record> records) {
         }
     }
 
+    /**
+     * A group can be upgraded offline if it's a classic group and empty.
+     *
+     * @param groupId The group to be validated.
+     * @return true if the offline upgrade is valid.
+     */
+    private boolean validateOfflineUpgrade(String groupId) {
+        Group group = groups.get(groupId);
+        return group != null && group.type() == CLASSIC && group.isEmpty();
+    }
+
+    /**
+     * A group can be downgraded offline if it's a consumer group and empty.
+     *
+     * @param groupId The group to be validated.
+     * @return true if the offline downgrade is valid.
+     */
+    private boolean validateOfflineDowngrade(String groupId) {
+        Group group = groups.get(groupId);
+        return group != null && group.type() == CONSUMER && group.isEmpty();
+    }
+
+    /**
+     * Upgrade/Downgrade the empty group if it's valid.
+     *
+     * @param groupId The group id to be migrated.
+     * @param records The list of records to delete the previous group.
+     */
+    public void maybeMigrateEmptyGroup(String groupId, List<Record> records, boolean isUpgrade) {
+        if ((isUpgrade && validateOfflineUpgrade(groupId)) ||
+            (!isUpgrade && validateOfflineDowngrade(groupId))) {
+            deleteGroup(groupId, records);
+            removeGroup(groupId);
```

Comment:
There is a bug for downgrade. The sequence of downgrading now is 1) consumer group gets deleted 2) classicGroupJoin creates a classic group and put it into `groupMetadataManager.groups`3) replaying consumer group tombstone from `deleteGroup` where `groups.get(groupId)` should be a consumer group.

Reply:
Do we actually replay records in this particular case? The replay handling for the classic protocol is a big different from all the other ones. See here: https://github.com/apache/kafka/pull/15442/files#diff-00f0f81cf13e66781777d94f7d2e68a581663385c37e98792507f2294c91bb09R1978.